include/tbbexec/tbb_thread_pool.hpp

```diff
@@ -0,0 +1,413 @@
+                return tbb_thread_pool::scheduler{s.pool_};
+            }
+
+            friend struct tbb_thread_pool::scheduler;
+
+            explicit sender(tbb_thread_pool& pool) noexcept : pool_(pool) {}
+
+            tbb_thread_pool& pool_;
+        };
+
+        template <class Fun, class Shape, class... Args>
+            requires stdexec::__callable<Fun, Shape, Args...>
+        using bulk_non_throwing = stdexec::__bool<
+            // If function invocation doesn't throw
+            stdexec::__nothrow_callable<Fun, Shape, Args...>&&
+            // and emplacing a tuple doesn't throw
+            noexcept(stdexec::__decayed_tuple<Args...>(std::declval<Args>()...))
+            // there's no need to advertise completion with `exception_ptr`
+            >;
+
+        template <class SenderId, class ReceiverId, class Shape, class Fun, bool MayThrow>
+        struct bulk_shared_state : exec::task_base {
+            using Sender = stdexec::__t<SenderId>;
+            using Receiver = stdexec::__t<ReceiverId>;
+
+            using variant_t = stdexec::__value_types_of_t<Sender,
+                                                          stdexec::env_of_t<Receiver>,
+                                                          stdexec::__q<stdexec::__decayed_tuple>,
+                                                          stdexec::__q<stdexec::__variant>>;
+
+            variant_t data_;
+            tbb_thread_pool& pool_;
+            Receiver receiver_;
+            Shape shape_;
+            Fun fn_;
+
+            std::atomic<std::uint32_t> finished_threads_{0};
+            std::atomic<std::uint32_t> thread_with_exception_{0};
+            std::exception_ptr exception_;
+
+            // Splits `n` into `size` chunks distributing `n % size` evenly between ranks.
+            // Returns `[begin, end)` range in `n` for a given `rank`.
+            // Example:
+            // ```cpp
+            // //         n_items  thread  n_threads
+            // even_share(     11,      0,         3); // -> [0,  4) -> 4 items
+            // even_share(     11,      1,         3); // -> [4,  8) -> 4 items
+            // even_share(     11,      2,         3); // -> [8, 11) -> 3 items
+            // ```
+            static std::pair<Shape, Shape> even_share(Shape n, std::uint32_t rank, std::uint32_t size) noexcept {
+                const auto avg_per_thread = n / size;
+                const auto n_big_share = avg_per_thread + 1;
+                const auto big_shares = n % size;
+                const auto is_big_share = rank < big_shares;
+                const auto begin =
+                    is_big_share ? n_big_share * rank : n_big_share * big_shares + (rank - big_shares) * avg_per_thread;
+                const auto end = begin + (is_big_share ? n_big_share : avg_per_thread);
+
+                return std::make_pair(begin, end);
+            }
```

Comment:
This is very similar to the version in `static_thread_pool.hpp`. Ideally the logic would be shared...